packages/snippet-manager/src/snippet-manager.ts

```diff
@@ -0,0 +1,527 @@
+      this.inflightFetchIndexPromise = null;
+
+      if (!fetchedFromNetwork && (Date.now() - (await fs.stat(cachePath)).mtime.getTime() > 3600_000)) {
+        // If the cache is old, we still use it, but we also kick off a new
+        // network fetch.
+        this.prepareIndex('force-refresh').catch(() => {});
+      }
+
+      return repoData;
+    })());
+  }
+
+  get snippets(): SnippetDescription[] {
+    return this.repos?.flatMap((repository: SnippetIndexFile) => repository.index) ?? [];
+  }
+
+  async registryBaseUrl(): Promise<string> {
+    return await this.config.get('snippetRegistryURL');
+  }
+
+  async ensureSetup(): Promise<string[]> {
+    if (this.npmArgv.length > 0 && this.repos) {
+      return this.npmArgv;
+    }
+
+    await this._internalState.interrupted.checkpoint();
+    [ this.npmArgv, this.repos ] = await Promise.all([
+      this.prepareNpm(),
+      this.prepareIndex()
+    ]);
+    return this.npmArgv;
+  }
+
+  async editPackageJSON<T>(fn: (pjson: any) => T): Promise<T> {
+    let pjson = {};
+    try {
+      pjson = JSON.parse(await fs.readFile(path.join(this.installdir, 'package.json'), 'utf8'));
+    } catch (err) {
+      this.messageBus.emit('mongosh-snippets:package-json-edit-error', { error: err.message });
+      if (err.code !== 'ENOENT') {
+        throw err;
+      }
+    }
+    const origContent = JSON.parse(JSON.stringify(pjson));
+    const result = await fn(pjson);
+    if (!isDeepStrictEqual(origContent, result)) {
+      await fs.writeFile(path.join(this.installdir, 'package.json'), JSON.stringify(pjson, null, '  '));
+    }
+    return result;
+  }
+
+  async runNpm(...npmArgs: string[]): Promise<string> {
+    await this.editPackageJSON(() => {}); // Ensure package.json exists.
+    return await this.execFile([
+      ...await this.ensureSetup(),
+      '--no-package-lock',
+      '--ignore-scripts',
+      `--registry=${await this.registryBaseUrl()}`,
+      ...npmArgs
+    ]);
+  }
+
+  async execFile([ cmd, ...args ]: string[]) {
+    const { interrupted } = this._internalState;
+    this.messageBus.emit('mongosh-snippets:spawn-child', { args: [ cmd, ...args ] });
+    await interrupted.checkpoint();
+    const proc = spawn(cmd, args, {
+      cwd: this.installdir,
+      env: { ...process.env, MONGOSH_RUN_NODE_SCRIPT: '1' },
+      stdio: 'pipe'
+    });
+    let stdout = '';
+    let stderr = '';
+    (proc.stdout as NodeJS.ReadableStream).setEncoding('utf8').on('data', (chunk) => { stdout += chunk; });
+    (proc.stderr as NodeJS.ReadableStream).setEncoding('utf8').on('data', (chunk) => { stderr += chunk; });
+
+    const interruptPromise = interrupted.asPromise();
+    interruptPromise.promise.catch(() => {}); // Suppress UnhandledPromiseRejectionWarning
+    try {
+      const [ exitCode ] = await Promise.race([
+        interruptPromise.promise,
+        once(proc, 'close')
+      ]);
+      if (exitCode === 0 || (exitCode === 1 && stderr === '' && stdout)) {
```

Comment:
So a non-zero exit code `1` is allowed if the program only prints to `stdout`? 🤔

Reply:
Err, right, this could use a comment – the important part is that `stderr` is empty:

```suggestion
      // Allow exit code 1 if stderr is empty, i.e. no error occurred, because
      // that is how commands like `npm outdated` report their result
      if (exitCode === 0 || (exitCode === 1 && stderr === '' && stdout)) {
```

Reply:
Ah alright, perfect! :) Having that comment would be really nice 👍